File: 2 Data Structures/Week 3 Priority Queues and Disjoint Sets/Programming-Assignment-2/make_heap/build_heap.py
# python3

class HeapBuilder:
  def __init__(self):
    self._swaps = []
    self._data = []

  def ReadData(self):
    global n
    n = int(input())
    self._data = [int(s) for s in input().split()]
    assert n == len(self._data)

  def WriteResponse(self):
    print(len(self._swaps))
    for swap in self._swaps:
      print(swap[0], swap[1])
  
  def parent(self, i):
      return (i - 1) // 2

  def LeftChild(self, i):
      return 2 * i + 1 
      
  def RightChild(self, i):
      return 2 * i + 2
      
  def ShiftDown(self, i):
      minIndex = i
      left = self.LeftChild(i)
      if left < n and self._data[left] < self._data[minIndex]:
          minIndex = left
      right = self.RightChild(i)
      if right < n and self._data[right] < self._data[minIndex]:
          minIndex = right
      if i != minIndex:
          self._data[i], self._data[minIndex] = self._data[minIndex], self._data[i]
          self._swaps.append([i, minIndex])
          self.ShiftDown(minIndex)

  def GenerateSwaps(self):
    # The following naive implementation just sorts 
    # the given sequence using selection sort algorithm
    # and saves the resulting sequence of swaps.
    # This turns the given array into a heap, 
    # but in the worst case gives a quadratic number of swaps.
    #
    # TODO: replace by a more efficient implementation
    for i in range(len(self._data)):
      for j in range(i + 1, len(self._data)):
        if self._data[i] > self._data[j]:
          self._swaps.append((i, j))
          self._data[i], self._data[j] = self._data[j], self._data[i]
    
  def QuickGenerateSwaps(self):
    for i in range(n // 2, -1, -1):
        self.ShiftDown(i)

  def Solve(self):
    self.ReadData()
    self.QuickGenerateSwaps()
    self.WriteResponse()

if __name__ == '__main__':
    heap_builder = HeapBuilder()
    heap_builder.Solve()

File: 2 Data Structures/Week 3 Priority Queues and Disjoint Sets/Programming-Assignment-2/make_heap/build_heap.cpp
#include <iostream>
#include <vector>
#include <algorithm>

using std::vector;
using std::cin;
using std::cout;
using std::swap;
using std::pair;
using std::make_pair;

using namespace std;

class HeapBuilder {
 private:
  vector<int> data_;
  vector< pair<int, int> > swaps_;

  void WriteResponse() const {
    cout << swaps_.size() << "\n";
    for (int i = 0; i < swaps_.size(); ++i) {
      cout << swaps_[i].first << " " << swaps_[i].second << "\n";
    }
  }

  void ReadData() {
    int n;
    cin >> n;
    data_.resize(n);
    for(int i = 0; i < n; ++i)
      cin >> data_[i];
  }

  void GenerateSwaps() {
    swaps_.clear();
    // The following naive implementation just sorts 
    // the given sequence using selection sort algorithm
    // and saves the resulting sequence of swaps.
    // This turns the given array into a heap, 
    // but in the worst case gives a quadratic number of swaps.
    //
    // TODO: replace by a more efficient implementation
    for (int i = 0; i < data_.size(); ++i)
      for (int j = i + 1; j < data_.size(); ++j) {
        if (data_[i] > data_[j]) {
          swap(data_[i], data_[j]);
          swaps_.push_back(make_pair(i, j));
        }
      }
  }
  
  int parent(int i){
      return (i - 1) / 2;
  }
  
  int LeftChild(int i){
      return 2 * i + 1;
  }
  
  int RightChild(int i){
      return 2 * i + 2;
  }
  
  void ShiftDown(int i){
      int n = data_.size();
      int minIndex = i;
      int left = LeftChild(i);
      if (left < n && data_[left] < data_[minIndex]){
          minIndex = left;
      }
      int right = RightChild(i);
      if (right < n && data_[right] < data_[minIndex]){
          minIndex = right;
      }
      if (i != minIndex){
          swap(data_[i], data_[minIndex]);
          swaps_.push_back(make_pair(i, minIndex));
          ShiftDown(minIndex);
      }
  }
  
  void QuickGenerateSwaps(){
      for (int i = data_.size() / 2; i > -1; i--){
          ShiftDown(i);
      }
  }

 public:
  void Solve() {
    ReadData();
    QuickGenerateSwaps();
    WriteResponse();
  }
};

int main() {
  std::ios_base::sync_with_stdio(false);
  HeapBuilder heap_builder;
  heap_builder.Solve();
  return 0;
}
